src/sinks/util/service.rs

```diff
@@ -1,35 +1,61 @@
-use std::{hash::Hash, marker::PhantomData, sync::Arc, time::Duration};
+use std::{hash::Hash, marker::PhantomData, pin::Pin, sync::Arc, time::Duration};
 
+use futures_util::{future::BoxFuture, Stream, TryStream, TryStreamExt};
 use serde::{Deserialize, Serialize};
 use tower::{
+    balance::p2c::Balance,
+    discover::Change,
     layer::{util::Stack, Layer},
     limit::RateLimit,
     retry::Retry,
-    timeout::Timeout,
+    timeout::{Timeout, TimeoutLayer},
     Service, ServiceBuilder,
 };
 use vector_buffers::Acker;
 
 pub use crate::sinks::util::service::{
+    clone::{CloneLayer, CloneService},
     concurrency::{concurrency_is_none, Concurrency},
+    health::HealthService,
     map::Map,
 };
-use crate::sinks::util::{
-    adaptive_concurrency::{
-        AdaptiveConcurrencyLimit, AdaptiveConcurrencyLimitLayer, AdaptiveConcurrencySettings,
+use crate::{
+    internal_events::OpenGauge,
+    sinks::util::{
+        adaptive_concurrency::{
+            AdaptiveConcurrencyLimit, AdaptiveConcurrencyLimitLayer, AdaptiveConcurrencySettings,
+        },
+        retries::{FixedRetryPolicy, RetryLogic},
+        service::map::MapLayer,
+        sink::Response,
+        Batch, BatchSink, Partition, PartitionBatchSink,
     },
-    retries::{FixedRetryPolicy, RetryLogic},
-    service::map::MapLayer,
-    sink::Response,
-    Batch, BatchSink, Partition, PartitionBatchSink,
 };
 
+mod clone;
 mod concurrency;
+mod health;
 mod map;
 
-pub type Svc<S, L> = RateLimit<AdaptiveConcurrencyLimit<Retry<FixedRetryPolicy<L>, Timeout<S>>, L>>;
+pub type InnerSvc<S, L> = RateLimit<AdaptiveConcurrencyLimit<Retry<FixedRetryPolicy<L>, S>, L>>;
+pub type Svc<S, L> = InnerSvc<Timeout<S>, L>;
 pub type TowerBatchedSink<S, B, RL> = BatchSink<Svc<S, RL>, B>;
 pub type TowerPartitionSink<S, B, RL, K> = PartitionBatchSink<Svc<S, RL>, B, K>;
+pub type DistributedService<S, RL, K, Req> = InnerSvc<
+    CloneService<
```

Comment:
What is the purpose of the clone layer here?

Reply:
The same reason ARC controller needs to be. At least retry layer requires cloneable service so that it can retry requests, `Balance` isn't cloneable, and all of the requests need to pass through the same `Balance`, so `CloneService` layer is needed.